### ISSUE TITLE
To do

### ISSUE THREAD

**Issue by llwyd** [llwyd/wavPlay#1]
Rather than having this in the readme file I think it would be better to have them here.  My ideas for features so far:

    * File Playlist (like winamp?)
    * More file types
    * Handle Stereo files
    * Improve user interface

**Comment by llwyd** [comment]
I think more file types can be achieved by using an encoder such as LAME (for mp3).

**Comment by xiewen112** [comment]
yes, using an encoder is the right way. 

    On Tuesday, March 8, 2016 11:35 AM, Tom <notifications@github.com> wrote:
 

 I think more file types can be achieved by using an encoder such as LAME (for mp3).—

**Comment by xiewen112** [comment]
Oops, I just commented on wrong issue. Extracting human voice maybe a good idea. We add a function to extract human voice from noisy surrounding or songs. That's must be cool and useful in many areas.

**Comment by llwyd** [comment]
I agree, I think it would be a pretty cool feature!  I also think having a general filtering system/equaliser would be cool too.  But for that I think we would have to restructure how the audio is being played so that it is being sent in buffers, rather than all at once.

**Comment by xiewen112** [comment]
I agree with it. I also think that we should draw a good structure for player, which must be high compatible, more explicit function dividing and robust. 

    On Tuesday, March 8, 2016 11:53 AM, Tom <notifications@github.com> wrote:
 

 I agree, I think it would be a pretty cool feature! I also think having a general filtering system/equaliser would be cool too. But for that I think we would have to restructure how the audio is being played so that it is being sent in buffers, rather than all at once.—

**Comment by llwyd** [comment]
I agree, I also think redesigning the user interface to accommodate the new features would be useful too.  I quite like the compactness of it so far though.

**Comment by xiewen112** [comment]
LOL, we are not professional UI designer. The program will be modified again and again, so I think we can complete these features first, and then begin to consider about the UI designing.  

    On Tuesday, March 8, 2016 12:13 PM, Tom <notifications@github.com> wrote:
 

 I agree, I also think redesigning the user interface to accommodate the new features would be useful too. I quite like the compactness of it so far though.—

**Comment by llwyd** [comment]
Aye but we can still make the layout look a little nicer!

**Comment by xiewen112** [comment]
yeah, let's do it

    On Tuesday, March 8, 2016 12:43 PM, Tom <notifications@github.com> wrote:
 

 Aye but we can still make the layout look a little nicer!—

**Comment by drakaru** [comment]
Getting buffered playback would be nice!

I can do ui quite well but haven't played around QT, I could take a look, see what I can come up with and bring some ideas to the table

**Comment by llwyd** [comment]
I have some buffering code working.  I need to test it more before committing it.

**Comment by xiewen112** [comment]
Pretty cool, looking forward to it. 

    On Friday, March 11, 2016 2:43 PM, Tom <notifications@github.com> wrote:
 

 I have some buffering code working. I need to test it more before committing it.—

**Comment by llwyd** [comment]
Took me a while but I have filtering working.

**Comment by llwyd** [comment]
Now for the task of putting all the code in the wavplayer D:

**Comment by llwyd** [comment]
I've created a wiki so that we can list all the things we need to do/bugs etc.

**Comment by llwyd** [comment]
I've been thinking about the cross platform aspect of thunderWave.  I think it might be worth exploring using something like PortAudio (http://www.portaudio.com/) where it can be built for different audio libraries such as ALSA on linux and WASAPI on windows etc.  I've got a small amount of experience with using it but I'd be willing to have a proper go at it for thunderWave.  Either that or I'd have to learn ALSA/pulseaudio from scratch, which I dont mind doing but it is a significantly larger learning curve.

**Comment by drakaru** [comment]
If there's a library to do it for us, why not? Reminds me of a principal of OOP: DRY (don't repeat yourself)
At the least, definitely worth looking at \o/

**Comment by llwyd** [comment]
On a first look Audacity and VLC media player both use it, so it must be legit :D

**Comment by llwyd** [comment]
:penguin: Turns out Github has emojis...

**Comment by drakaru** [comment]
If it's good enough for VLC, it's good enough for me! :penguin:

**Comment by llwyd** [comment]
I concur!  I will get to work on implementing it >:)

**Comment by llwyd** [comment]
It has been a while since I posted here.  There have been some minor updates.

The first is that I've successfully got the portaudio library working and I'm slowly implementing it into the existing program.

The second is that I'm thinking of forking the project.  I'm going to keep thunderWave as a pure real time spectrum analyser (Which is what it is doing at the moment) and remove features such as the equaliser etc.  Instead I'm creating a new project which is called "uWave" (microwave, ha) which will be a cross platform lightweight audioplayer, but this will lack the graphical analysers.

**Comment by llwyd** [comment]
I've made some significant developments using portaudio:

mono AND stereo playback streaming from a file.  I'm very happy I've got this working.  Now to try and implement it in the program.  Interestingly I've got portaudio to work with QT fine in linux but not in windows >:(

**Comment by xiewen112** [comment]
Wow, pretty cool. I always think that it is hard to operate when you want to use some hardware device on Windows. But, anyway, we may can finally solve the problem on Windows.
